core/src/main/kotlin/com/strumenta/kolasu/ids/DeclarativeNodeIdProvider.kt

```diff
@@ -0,0 +1,78 @@
+package com.strumenta.kolasu.ids
+
+import com.strumenta.kolasu.model.Node
+import kotlin.reflect.KClass
+import kotlin.reflect.full.isSubclassOf
+import kotlin.reflect.full.isSuperclassOf
+
+/**
+ * Declarative node id provider instances allow to define
+ * language-specific or custom identifier construction policies.
+ *
+ * ```kotlin
+ * object MyDeclarativeNodeIdProvider: DeclarativeNodeIdProvider(
+ *     idFor(PackageDeclaration::class) { it.node.name }
+ *     idFor(ClassDeclaration::class) { (node: ClassDeclaration, identifiers: NodeIdProvider) ->
+ *         val packageIdentifier = node.findAncestorOfType(PackageDeclaration::class)?.let { identifiers.idFor(it) } ?: "__none__"
+ *         "${packageIdentifier}::${node.name}"
+ *     }
+ * )
+ * ```
+ **/
+open class DeclarativeNodeIdProvider(
+    vararg rules: DeclarativeNodeIdProviderRule<out Node>
+) : NodeIdProvider {
+    private val rules: List<DeclarativeNodeIdProviderRule<out Node>> = rules.sorted()
+
+    override fun id(kNode: Node): String {
+        return this.rules.firstOrNull { it.canBeInvokedWith(kNode::class) }?.invoke(this, kNode)
+            ?: throw RuntimeException("Cannot find rule for node type: ${kNode::class.qualifiedName}")
+    }
+}
+
+/**
+ * Utility function to define declarative node id provider rules.
+ * Can be used whenever listing the rules in the DeclarativeNodeIdProvider constructor.
+ **/
+inline fun <reified NodeTy : Node> idFor(
+    noinline specification: (DeclarativeNodeIdProviderRuleContext<NodeTy>) -> String
+): DeclarativeNodeIdProviderRule<NodeTy> = DeclarativeNodeIdProviderRule(NodeTy::class, specification)
+
+/**
+ * Class representing a single rule of a DeclarativeNodeIdProvider.
+ **/
+class DeclarativeNodeIdProviderRule<NodeTy : Node>(
```

Comment:
If this class is never instantiated directly but always through `idFor`, could it make sense to make it private?

Reply:
It would produce an error as the `DeclarativeNodeIdProvider` takes instances of this class as input. The `idFor` is just an inlined utility method to build instances of these, which therefore need  to be publicly accessible.

Reply:
Then could we make its constructor private or internal, to communicate that people should not use it directly?

Reply:
It would still produce an error as `idFor` is inlined. This function is just syntactic sugar, users could actually use the `DeclarativeNodeIdProviderRule` constructor if they want. We would not be exposing internals anyway.

Reply:
Ok, marking this as resolved